### DISCUSSION TITLE
Practice for using type annotations in Airflow components

### DISCUSSION THREAD

**Discussion by pykenny** [apache/airflow#34155] [Q&A]
I'm currently replacing operator class instances with called decorated operators in some managed DAGs, and defining tasks with style like the DAG definition example below (under Airflow 2.5 and Python 3.8):

```python
from typing import TYPE_CHECKING
from airflow.decorators import dag, task

# ...

# Only introduce type on editors, linters, etc.
if TYPE_CHECKING:
    from typing import List, Optional

    from airflow.models import TaskInstance

# ...

CAPACITY_SPACE_MOUNTAIN = 6

@task(task_id="boarding")
def boarding() -> "List[str]":
   from faker import Faker

   rand_data_generator = Faker()

   return [rand_data_generator.name() for _ in range(CAPACITY_SPACE_MOUNTAIN)]

@task(task_id="log_exclaims")
def log_exclaims(passengers: "List[str]", ti: "Optional[TaskInstance]" = None):
    for passenger_name in passengers:
        ti.log.info(f"{passenger_name}: Whoo-hoo!")

# ...

@dag(dag_id: "space_mountain_ride", schedule_interval="*/10 * * * *")
def space_mountain_ride():
    boarding_task = boarding()
    exclaims_task = log_exclaims()

    _ = boarding_task >> exclaims_task
   

dag = space_mountain_ride()
```

Expected behaviors:
 - Annotations should be omitted during runtime
 - Annotations will be identified and processed by editors, linters, etc. when `TYPE_CHECKING` flag is on.

And what happens actually:
 - Error raised when scheduler loads the DAG:
   ```plaintext
   Broken DAG: [...] Traceback (most recent call last):
     File "/.../.../typing.py", line 518, in _evaluate
       eval(self.__forward_code__, globalns, localns),
     File "<string>", line 1, in <module>
   NameError: name 'List' is not defined
   ```

Of course, if the introduced type classes are imported anyway, the scheduler will be happy:

```python
from typing import List, Optional
from airflow.decorators import dag, task
from airflow.models import TaskInstance


# ...

CAPACITY_SPACE_MOUNTAIN = 6

@task(task_id="boarding")
def boarding() -> "List[str]":
   from faker import Faker

   rand_data_generator = Faker()

   return [rand_data_generator.name() for _ in range(CAPACITY_SPACE_MOUNTAIN)]

@task(task_id="log_exclaims")
def log_exclaims(passengers: "List[str]", ti: "Optional[TaskInstance]" = None):
    for passenger_name in passengers:
        ti.log.info(f"{passenger_name}: Whoo-hoo!")

# ...

@dag(dag_id: "space_mountain_ride", schedule_interval="*/10 * * * *")
def space_mountain_ride():
    boarding_task = boarding()
    exclaims_task = log_exclaims()

    _ = boarding_task >> exclaims_task
   

dag = space_mountain_ride()
```

Didn't check the internals but I guessed Airflow somewhat inspect type annotations in runtime and that caused the error in the former implementation. That makes me wonder if there's any suggested way/practice to write type annotations/hints in decorated operators and standard operator classes (e.g. there's no equivalent decorator for `ExternalTaskSensor` and `BranchDateTimeOperator` so you may need to include both type of components together in a DAG). Discussions/Issues such as #8996 and #19884 show that Airflow does use annotations for some advanced features and plans to take a step further as Python's type hinting system gets more robust (I use type annotations mostly just for quick intellisense and linting on editors 😅), but did not find out much about best practices.

**Comment by Taragolis** [comment]
I'm not sure that Airflow check annotations in runtime, at least I do not see that in core, but I could miss something.
Anyway, I've check your sample and with minor changes it work greats in Airflow 2.6.3 on Python 3.9 (I've just have it for other checks)

<details>
  <summary>Code under the spoiler</summary>

```python
from typing import TYPE_CHECKING
from airflow.decorators import dag, task
from airflow.utils.timezone import datetime


if TYPE_CHECKING:
    from typing import List, Optional

    from airflow.models import TaskInstance

CAPACITY_SPACE_MOUNTAIN = 6


@task(task_id="boarding")
def boarding() -> "List[str]":
    from faker import Faker

    rand_data_generator = Faker()

    return [rand_data_generator.name() for _ in range(CAPACITY_SPACE_MOUNTAIN)]


@task(task_id="log_exclaims")
def log_exclaims(passengers: "List[str]", ti: "Optional[TaskInstance]" = None):
    for passenger_name in passengers:
        ti.log.info(f"{passenger_name}: Whoo-hoo!")


@dag(dag_id="space_mountain_ride", schedule="*/10 * * * *", start_date=datetime(2023, 9, 1), catchup=False)
def space_mountain_ride():
    log_exclaims(boarding())


dag = space_mountain_ride()
```
  
</details>

There is some nitpicks
1. Personally I'm not sure there is any benefits to move `from typing import List, Optional` under `TYPE_CHECKING` block, it imports in the same module, so this objects already imported.
2. You could add `from __future__ import annotations` and use Python 3.9/3.10 annotations in Python 3.8, see: https://docs.python.org/3.8/library/__future__.html

<details>
  <summary>Python 3.10+ styles annotations</summary>

```python
from __future__ import annotations

from typing import TYPE_CHECKING
from airflow.decorators import dag, task
from airflow.utils.timezone import datetime


if TYPE_CHECKING:
    from airflow.models import TaskInstance


CAPACITY_SPACE_MOUNTAIN = 6


@task(task_id="boarding")
def boarding() -> list[str]:
    from faker import Faker

    rand_data_generator = Faker()

    return [rand_data_generator.name() for _ in range(CAPACITY_SPACE_MOUNTAIN)]


@task(task_id="log_exclaims")
def log_exclaims(passengers: list[str], ti: TaskInstance | None = None):
    for passenger_name in passengers:
        ti.log.info(f"{passenger_name}: Whoo-hoo!")


@dag(dag_id="space_mountain_ride", schedule="*/10 * * * *", start_date=datetime(2023, 9, 1), catchup=False)
def space_mountain_ride():
    log_exclaims(boarding())


dag = space_mountain_ride()
```
</details>

**Reply by pykenny** [reply]
1. `typing` load all the ~~names~~ definitions of built-in types when it gets imported so I'm fine with importing built-in types explicitly. It's just personal habit/decision to make all the names that are only used for type annotation go into the `TYPE_CHECKING` part.
2. The "3.10+ style annotations" example will not pass without importing `__future__.annotations` even under Python 3.10+, because `TaskInstance` still can't be found during runtime unless turning on type-checking flag, or using `__future__.annotations`.
   
   Sorta minimum example (swap `TaskInstance` class with some Python's built-in type name for fast implementation):
   
   ```python
   from typing import TYPE_CHECKING
   
   if TYPE_CHECKING:
       from typing import List
   
   
   def print_int_list(ls: List[int] | None = None):
       print(','.join(str(v) for v in (ls or [3, 4, 5])))
   
   print_int_list([1, 2, 3])
   ```
   Error (under Python 3.10.8):
   ```plaintext
   Traceback (most recent call last):
     File "/.../typing_example.py", line 7, in <module>
       def print_int_list(ls: List[int] | None = None):
   NameError: name 'List' is not defined. Did you mean: 'list'?
   ```
   [Root cause](https://docs.python.org/3/library/__future__.html#id1).

4. If "2.6.3 + Py 3.9" combination doesn't quirk in this case, then updating package/environment seems to be the fastest solution, given that both Python and Airflow update themselves frequently. (Running into some tangling dependencies issues in the pipeline codebase now, so it might take longer to start off the migration)

**Reply by potiuk** [reply]
> 1. typing load all the names definitions of built-in types when it gets imported so I'm fine with importing built-in types explicitly. It's just personal habit/decision to make all the names that are only used for type annotation go into the TYPE_CHECKING part.

I think it's not the best habit (not consistent with the idea of `TYPE_CHECKING` and non forward-looking). While originally type-hinting was meant to not be used to perform the actual type checking, stripping the typing information from runtime will break a number of libraries that use them. For example Pydantic - which is widely popular serialization framework, uses runtime type-hinting information to validate and serialize Python objects. Many popular frameworks adopted it (fast-api for example) and there are also other ways how type-checking at runtime can be used. For example `black` got some 10x faster after using mypyc and being fully type-hinted and mypyc uses type-hinting at runtime to optimize the code significantly.  

Seems that Python world generally recognized it and after the initial move to add typhinting as runtime-available and makes use of it liberally. `TYPE_CHECKING`  is really there for "expensive imports" and "avoiding runtime cycles" see https://peps.python.org/pep-0563/#runtime-annotation-resolution-and-type-checking - using it for "all type hints" is simply not the intended use for it and as noted in the PEP:

> Trying to resolve annotations of a_func at runtime using typing.get_type_hints() will fail since the
name expensive_mod is not defined (TYPE_CHECKING variable being False at runtime).
This was already true before this PEP.

So by using TYPE_CHECKING everywhere, you LITERALLY (pun intended) strips functionality that is already used by a number of libraries out there.

**Reply by pykenny** [reply]
I see. So seems that `TYPE_CHECKING` flag was invented as the last resort when hitting cases like circular references before `__future__.annotations` comes to the aid. I took the naming too literally 🤷‍♂️

**Reply by pykenny** [reply]
Back to the original question (as title) and outside of general practices in Python program development, it would be helpful to get examples about using type annotations (and perhaps as much as possible) in Airflow's DAG pipeline and task definitions, especially when a component requires accessing Airflow-specific context such as `ti`, `**context`, etc., either from documentation or community discussion.

**Reply by potiuk** [reply]
> it would be helpful to get examples about using type annotations

Absolutely @pykenny I am all for it. 

And after all those discussions and explanations you are - I think the best person to contribute such documentation. And it's super-easy - just find the right page where you would be looking for it in our docs, click "Suggest a change on this page" and PR will be opened where you will be able to contribute the docs.  Without leaving GitHub UI. And you will get reviews and comments from maintainers - but since you are passionate and knowledgeable about it, I think driving such a doc update is something you will be best at. Especially that you will write it from the users's perspective - without assumptions we have.  It is a very valuable contribution - you might join ~2600 people who created airflow (most of them - users like you who saw something missing and added it back) . And it's cool to help other users like you to find this information.

**Reply by Taragolis** [reply]
> The "3.10+ style annotations" example will not pass without importing __future__.annotations even under Python 3.10+, because TaskInstance still can't be found during runtime unless turning on type-checking flag, or using __future__.annotations.

Actual 3.10+ styles (without any additional imports)
```python
def print_int_list(ls: list[int] | None = None):
    print(','.join(str(v) for v in (ls or [3, 4, 5])))


print_int_list([1, 2, 3])
```

**Reply by Taragolis** [reply]
In general there is no recommendation for typing annotations. So personal recommendation would be use [`ruff`](https://github.com/astral-sh/ruff) in your project which might help to resolve different cases and even move into `TYPE_CHECKING` block imports which are used only in annotations, see [TCH rules](https://beta.ruff.rs/docs/rules/#flake8-type-checking-tch)
